akka-http-core/src/main/scala/akka/http/scaladsl/ClientTransportWithCustomResolver.scala

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.http.scaladsl
+
+import java.net.InetSocketAddress
+
+import akka.actor.ActorSystem
+import akka.annotation.ApiMayChange
+import akka.http.scaladsl.Http.OutgoingConnection
+import akka.http.scaladsl.settings.ClientConnectionSettings
+import akka.stream.scaladsl.{ Flow, Tcp }
+import akka.util.ByteString
+
+import scala.concurrent.{ ExecutionContext, Future }
+
+@ApiMayChange
+trait ClientTransportWithCustomResolver extends ClientTransport {
+  override def connectTo(host: String, port: Int, settings: ClientConnectionSettings)(implicit unused: ActorSystem): Flow[ByteString, ByteString, Future[Http.OutgoingConnection]] = {
+    // delay hostname resolution until stream materialization
+    Flow.setup { (mat, _) =>
+      implicit val ec: ExecutionContext = mat.executionContext
+      futureFlow {
+        inetSocketAddress(host, port, settings).map { address =>
+          Tcp()(mat.system).outgoingConnection(address, settings.localAddress,
+            settings.socketOptions, halfClose = true, settings.connectingTimeout, settings.idleTimeout)
+            .mapMaterializedValue(_.map(tcpConn => OutgoingConnection(tcpConn.localAddress, tcpConn.remoteAddress)))
+        }
+      }
+    }.mapMaterializedValue(_.flatten.flatten)
```

Comment:
The heart of this PR. But the doubly deferred construction looks like a monster.